### PR TITLE
9853 hot hod map storage single use objects

### LINKED ISSUE #9853: Hot Rod map storage: Single-use token (action token, code2token) no-downtime store
### Description

Create a JPA map storage implementation for single-use tokens

### PR DESCRIPTION
**Pull request by martin-kanis**
Closes #9853

### CODE REVIEW

**Review by mhajas — COMMENTED** [review]
Thanks for the PR @martin-kanis. I added one comment.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/SingleUseObjectHotRodMapStorage.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+        if (criteria == null) {
+            return Stream.empty();
+        }
+
+        SingleUseObjectModelCriteriaBuilder mcb = criteria.flashToModelCriteriaBuilder(createSingleUseObjectCriteriaBuilder());
+        if (mcb.isValid()) {
+            HotRodSingleUseObjectEntityDelegate value = read(mcb.getKey());
+
+            return value.getHotRodEntity() != null ? Stream.of(value) : Stream.empty();
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/SingleUseObjectHotRodMapStorage.java]: I kept the condition for now and created issue to fix HotRodMapStorage read/update methods https://github.com/keycloak/keycloak/issues/12304

[the remark this answers is not in the document]
Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/SingleUseObjectHotRodMapStorage.java]: Thanks @martin-kanis

**Review by mhajas — APPROVED** [review]
Thanks, LGTM

**Review by hmlnarik — APPROVED** [review]
Approving per @mhajas 's review

**Pull request merged**